src/main/java/com/zendesk/maxwell/MaxwellAbstractRowsEvent.java

```diff
@@ -275,21 +279,75 @@ public Object getData(String string) {
 		return list;
 	}
 
-	public List<JSONObject> toJSONObjects() {
-		ArrayList<JSONObject> list = new ArrayList<>();
+	static final JsonFactory factory = new JsonFactory();
 
-		for ( RowMap map : jsonMaps() ) {
-			list.add(new MaxwellJSONObject(map));
+	private JsonGenerator createJSONGenerator(ByteArrayOutputStream b) {
+		try {
+			return factory.createGenerator(b);
+		} catch (IOException e) {
+			LOGGER.error("Caught exception while creating JSON generator: " + e);
 		}
-		return list;
+		return null;
+	}
+
+	static String[] keyOrder = {"database", "table", "type", "ts"};
+
+	private void rowMapToJSON(JsonGenerator g, RowMap map) throws IOException {
+		g.writeStartObject(); // {
+
+		for ( String key: keyOrder ) {
+			g.writeObjectField(key, map.get(key)); // type: "insert"
+		}
+
+		if ( map.containsKey("xid") )
+			g.writeObjectField("xid", map.get("xid"));
+
+		if ( map.containsKey("commit") && (boolean) map.get("commit") == true)
+			g.writeBooleanField("commit", true);
+
+		g.writeObjectFieldStart("data");
+		for ( String key: map.data.keySet() ) {
+			Object data = map.getData(key);
+
+			if ( data == null )
+				continue;
+
+			if ( data instanceof List ) { // sets come back from .asJSON as lists, and jackson can't deal.
+				List<String> stringList = (List<String>) data;
+
+				g.writeArrayFieldStart(key);
+				for ( String s : stringList )  {
+					g.writeString(s);
+				}
+				g.writeEndArray();
+			} else {
+				g.writeObjectField(key, data);
+			}
+		}
+		g.writeEndObject(); // end of 'data: { }'
+
+		g.writeEndObject();
 	}
 
 	public List<String> toJSONStrings() {
 		ArrayList<String> list = new ArrayList<>();
+		ByteArrayOutputStream b = new ByteArrayOutputStream();
+		JsonGenerator g = createJSONGenerator(b);
+
+		g.setRootValueSeparator(null);
 
 		for ( RowMap map : jsonMaps() ) {
-			list.add(new MaxwellJSONObject(map).toString());
+			try {
+				rowMapToJSON(g, map);
+				b.reset();
+				g.flush();
+				list.add(b.toString());
+			} catch ( IOException e ) {
+				LOGGER.error("Caught IOException while generating JSON: " + e);
```

Comment:
Can do `LOGGER.error("Caught IOException while generating JSON: " + e, e);` instead of the `printStackTrace`